# UN-CNVc: the Unimaginatively-Named Copy Number Variant Caller

> We would like to thank the United Nations for serendipitously lending their name to our caller.

## Rationale
Unlike all of the CNV callers we know, `UN-CNVc` uses VCFs to call copy number variants. It requires high-depth (>10x, >20x is ideal) whole genome sequencing on at least 100 samples to work well. This input format makes it notably faster than alignment-based callers: it can process approximately 1 sample per second on a 10Mb chunk. This means that if you have access to ~150 processing cores for parallelisation, `un-cnvc` can process **1,000 samples in about 15 minutes**.

## Limitations
Due to the way it works, `un-cnvc` will not be able to genotype small events. Typically, sensitivity drops dramatically for events below 5-10kb in size, or events that are spanned by less than 50 SNVs.
Not only are small events more likely to be missed, they are also more likely to get genotyped wrong. The program uses two different genotyping algorithms, which both perform best in different cases. To help you make the decision, it generates a QC file which can help you decide which method to use. 

## Prerequisites
To run UN-CNVc, you will need :
* a recent (>v. 1.1) version of `bcftools`
* a recent (>v. 3.3) version of `R`
* the `mixtools`, `zoo`, `rpart` and `data.table` R libraries (`install.packages("mixtools")`, etc.)

## Installation
`un-cnvc` should work almost out-of-the-box: after checking out the file and making sure you have access to all dependencies above, please change the first line of `cnvc.R` to reflect the path of your local R interpreter.

## Usage

In version 1.1 the running process has been greatly simplified. First, generate per-chromosome depth files using `bcftools`. For example:

```bash
for i in {1..22}; do bcftools stats -s- chr$i.vcf.gz | grep ^PSC | awk '{print $3, $10}' > chr$i.avgdepth; done
```

After that, only 1 call is necessary:

```bash
./uncnvc --vcf [vcf.gz_file] --interval [interval_to_call] --depth-file [depth_file] --out [output_basename] --cp [complexity] --include-dups [include_dups] --method [method] [--no-genotypes]
```
* Options:
   * `-v|--vcf| [vcf.gz_file]` is a **bgzipped, tabixed VCF file** which contains `[interval_to_call]`. Currently only chromosome names prefixed with `chr` are supported.
   * `-i|--interval [interval_to_call]` is the genomic interval you wish to call depth in (we strongly advise using 10Mb windows). Can be written with or without the leading `chr`.
   * `-d|--depth-file [depth_file]` is the depth file generated by the call above. It is a headerless file, with two columns: sample ID and chromosome-wide average depth.
   * `-o|--out [output_basename]` is the base output name for all files.
   * `-c|--complexity [complexity]` is the complexity parameter passed to `rpart`. In practice, a value between 0.05 and 0.001 gives best results; please see below for how to tune this value.
   * `-u|--include-dups [0|1]` controls whether duplication-only events are called.
      * `0` calls only deletions and calls duplications the same as no CNV. This generates a PLINK `.ped/.map` dataset.
      * `1` calls duplications as well as deletions. This generates a PLINK CNV dataset (`.cnv`).
   * `-m|--method [segment|means|both]` is the genotyping method used. See below for a discussion. We advise to use `both`. Ignored if `-s` is set.
   * `-s|--no-genotypes` suppresses genotyping output, just produces diagnostics plots. Useful for tuning the `-c` parameter.

## Output
* `[output_filename].[interval_to_call].alldp`. A (large) file which contains depth measurement for all samples at all markers in the region.
* `[output_filename].[interval_to_call].window.pdf`. A graph such as the one below:
![example.png](example.png)
   * The top left panel represents the piecewise constant relative depth intervals. There will be a cluster around 1, representing the normal depth. Horizontal dashed intervals represent expected locations of CNV segments (het/hom del or dup). Vertical highlighted regions are regions called as variants by UN-CNVc.
   * Top right panel is a histogram of the observed segment depths.
   * Bottom left panel represents the statistics used by the caller to call variable regions.
* `[output_filename].ped/.map` PLINK dataset. Produced when the duplicate flag is set to 0.
* `[output_filename].cnv` PLINK CNV dataset. Produced when the duplicate flag is set to 1. Currently PLINK support for this format is in beta status.
* `[output_filename].[interval_to_call].QC.pdf`. A PDF file containing for each called event in the interval, a graph such as the one below:
![example_QC.png](example_QC.png)
   * The top left panel represents average depth calls in the region, with depth calls shown as colors. A good graph has clearly defined clusters around multiples of 0.5 that each mostly contain 1 colour.
   * The top left panel shows rolling mean depths, per sample, in the region. Lines should be mostly horizontal (no slopes), and should be clustered around multiples of 0.5.
   * The bottom right panel represents piecewise constant intervals, with samples containing multiple intervals in the region coloured in red. Check that most non-REF intervals are about the same length and largely overlap.
   * The bottom left panel is a diagnostics plot comparing genotyping results from both methods. Ideally, results will be similar, meaning that `n` clusters of mostly identical colors and shapes should appear on the diagonal of the plot. For small events this will not be the case, though, and this plot should help you decide which method to use, or whether to discard this SV as part of your QC pipeline.

A file with the following header, used to filter your CNVs:
   * `chr` 
   * `start`
   * `stop`
   * `hom` : number of homozygous deletions (depth=0)
   * `het` : number of heterozygous deletions (depth=0.5)
   * `normal` : number of samples with no variation
   * `dup` : number of samples with duplications (depth=k.0.5, with k>2)
   * `del_af` : frequency of deletion
   * `del_ratio` : ratio of deletions to duplications
   * `avg_del_p` : average confidence of the deletion calls
   * `del_hc` : number of high-confidence calls
   * `del_lc` : number of low-confidence calls (`confidence=5%` by default so low-confidence is really low confidence.)
   * `del_hc_total_ratio` : fraction of dels that are high-confidence 
   * `avg_dup_p` : average confidence of duplications
   * `dup_hc` : number of high-confidence duplication calls
   * `dup_lc`  : number of low-confidence dup calls
   * `dup_hc_total_ratio` : fraction of dups that are high-confidence
* `[output_filename].pdf` A diagnostics plot, used to visualise your results.

